modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/SearchProcessor.java

```diff
@@ -0,0 +1,278 @@
+     *     <li>password: elasticsearch password (for plain text auth)
+     *     <li>index: index we want to run the query on
+     *     <li>type: type we want to run the query on
+     *     <li>query: elasticsearch query to run, defaults to { "query" : { "match_all": {} } }
+     *     <li>path: defines a path to apply on the response to extract the object to be merged in the target field.
+     *         Defaults to hits.hits.0._source
+     *     <li>target_field: destination field. Defaults to "target".
+     * </ul>
+     */
+    public static final class Factory implements Closeable, Processor.Factory {
+
+        public static final String DEFAULT_HOST = "http://127.0.0.1:9200";
+        public static final String DEFAULT_QUERY = "{ \"query\" : { \"match_all\": {} } }";
+        public static final String DEFAULT_PATH = "hits.hits._source";
+
+        protected final Map<String, RestClient> clients;
+
+        public Factory() {
+            clients = new HashMap<>();
+        }
+
+        @Override
+        public SearchProcessor create(Map<String, Processor.Factory> registry,
+                                      String processorTag,
+                                      Map<String, Object> config) throws Exception {
+            // TODO we should default to the current node http layer if available
+            String host = readStringProperty(TYPE, processorTag, config, "host", DEFAULT_HOST);
+            String username = readOptionalStringProperty(TYPE, processorTag, config, "username");
+            // TODO probably a bad idea to have that in clear
+            String password = readOptionalStringProperty(TYPE, processorTag, config, "password");
+            String index = readOptionalStringProperty(TYPE, processorTag, config, "index");
+            String type = readOptionalStringProperty(TYPE, processorTag, config, "type");
+            String query = readStringProperty(TYPE, processorTag, config, "query", DEFAULT_QUERY);
+            String path = readStringProperty(TYPE, processorTag, config, "path", DEFAULT_PATH);
+            String targetField = readOptionalStringProperty(TYPE, processorTag, config, "target_field");
+
+            if (index != null && index.isEmpty()) {
+                throw newConfigurationException(TYPE, processorTag, "index", "when set can't be empty");
+            }
+
+            if (type != null && type.isEmpty()) {
+                throw newConfigurationException(TYPE, processorTag, "type", "when set can't be empty");
+            }
+
+            // If type is defined, index must be set
+            if (type != null && index == null) {
+                throw newConfigurationException(TYPE, processorTag, "index/type", ": if [type] is set, [index] must be set as well");
+            }
+
+            // We reuse the same connection to the same server instead of building a new instance every time
+            RestClient client = clients.get(host);
```

Comment:
It is but I agree that if we don't synchronize we could end up with 2 instances created for the same endpoint.